src/smt/theory_str_noodler/regex.cpp

```diff
@@ -0,0 +1,559 @@
+            } else {
+                util::throw_error("loop should contain at least lower bound");
+            }
+
+            // min_length: low == 0 --> 0; otherwise min_length * low
+            // empty: low == 0 --> false; otherwise the same as the original empty
+            // universal: min_length --> false; low == 0 --> false
+            RegexInfo res = get_regex_info(to_app(body), m_util_s, m);
+            if(res.empty == l_true && low == 0) {
+                return RegexInfo{.min_length = 0, .universal = l_false, .empty = l_false};
+            }
+            res.min_length *= low;
+            if(res.min_length > 0) {
+                res.universal = l_false;
+            }
+            return res;
+
+        } else if (m_util_s.re.is_of_pred(expression)) { // Handle of predicate.
+            util::throw_error("of predicate is unsupported");
+        } else if (m_util_s.re.is_opt(expression)) { // Handle optional.
+            SASSERT(expression->get_num_args() == 1);
+            const auto child{ expression->get_arg(0) };
+            SASSERT(is_app(child));
+            // min_length: 0 (epsilon)
+            RegexInfo res = get_regex_info(to_app(child), m_util_s, m);
+            res.min_length = 0;
+            return res;
+        } else if (m_util_s.re.is_range(expression)) { // Handle range.
+            SASSERT(expression->get_num_args() == 2);
+            const auto range_begin{ expression->get_arg(0) };
+            const auto range_end{ expression->get_arg(1) };
+            SASSERT(is_app(range_begin));
+            SASSERT(is_app(range_end));
+            const auto range_begin_value{ to_app(range_begin)->get_parameter(0).get_zstring()[0] };
+            const auto range_end_value{ to_app(range_end)->get_parameter(0).get_zstring()[0] };
+
+            // min_length: if there is some symbol in range --> min_length = 1; otherwise min_length = 0 (empty)
+            // empty:  if there is some symbol in range --> false; otherwise true
+            // universal: false
+            if(range_begin_value <= range_end_value) {
+                return RegexInfo{.min_length = 1, .universal = l_false, .empty = l_false};
+            } else {
+                return RegexInfo{.min_length = 0, .universal = l_false, .empty = l_true};
+            }
+        } else if (m_util_s.re.is_reverse(expression)) { // Handle reverse.
+            util::throw_error("reverse is unsupported");
+        } else if (m_util_s.re.is_union(expression)) { // Handle union (= or; A|B).
+            SASSERT(expression->get_num_args() == 2);
+            const auto left{ expression->get_arg(0) };
+            const auto right{ expression->get_arg(1) };
+            SASSERT(is_app(left));
+            SASSERT(is_app(right));
+            
+            // min_length: minimum of min_length of both guys
+            // empty: one of them is empty --> true; otherwise undef
+            // universal: min_length > 0 --> false; otherwise undef
```

Comment:
I think we can also add check that if one of them is universal -> true

Reply:
fixed